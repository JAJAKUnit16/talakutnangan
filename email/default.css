* {
	box-sizing: border-box;
}

:root, html, body {
	font-size: 16px;
	height: 100%;
	width: 100%;
}

body {
	display: flex;
	flex: row wrap;
	justify-content: center;
	align-items: center;
}

body > div {
	flex: 1 0 100%;
	margin: 1em 0.5em;
}


h1 { text-align: center; }
