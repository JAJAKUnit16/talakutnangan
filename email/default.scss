body {
	flex: 1 0 100%;
	margin: 10em 0.5em;
	background-color: #ebf4f5
}


.box-column {
	margin: auto;
	width: 100%;
	max-width: 640px;
	padding: 20px;
	border-radius: 25px;
	background-color: white;
  }

.content{
	text-align: center;
	padding: 10px;

	.home-page-url::before {
		content: " ";
	}
}

.pass{
	color:#1e4eee;
	font-family: Impact;
	font-size: 180%;
	text-align: center;

}

.buttons{
	border: none;
	padding: 5px 5px;

	width: 100px;
	height: max-content;

	background-color: #4CAF50;
	color: white;
	cursor: pointer;
	font-size: 16px;
	text-align: center;
	text-decoration: none;
  }

.line{
	@apply flex justify-center items-center;

	margin-bottom: 1em;
	border-bottom: 2px solid #b0a5a5;
	padding-bottom: 1em;
}

.note {
	text-align: left;
	padding: 10px;
	color: #b0a5a5;
}

h1 { 
	text-align: center;
	font-family:Impact; 
	padding: 20px;
	font-size: 200%;
}




