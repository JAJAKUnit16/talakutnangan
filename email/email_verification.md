<template>
<div class="box-column">
<h1>Email Verification</h1>

<p class="content">
    Your e-mail ({{ email }}) has been registered in <a class="home-page-url" href="{{ homePageURL }}">Talakutnangan.</a>
</p>

<p class ="content">Please click the button below to verify:</p>
<div class="line">
    <div class="buttons">
        <button type="button" href="{{ emailVerificationURL }}">Verify e-mail</button>
    </div>
</div>
<p class="note">Note</p>
<p class="note">If the button above does not work properly, visit {{ emailVerificationURL }}</p>
</div>

</template>