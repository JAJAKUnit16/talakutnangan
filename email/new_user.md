<div class="box-column">
<h1>New User</h1>

<p class="content">
    Hello {{ name }}, your e-mail address ({{ email }}) have been registered as a {{ kind }} in <a class="home-page-url" href="{{ homePageURL }}">Talakutnangan.</a>
</p>

<p class ="content">Please log in using the following password:</p>

<div class="line">
<p class="pass">{{ password }}</p>
</div>

<p class="note">Note</p>
<p class="note">Upon logging in, do not share the password and change the default password.</p>
</div>
