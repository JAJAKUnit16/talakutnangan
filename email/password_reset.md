# Password Reset

Hello {{ name }} ({{ email }}), your password has been reset by the admin in [Talakutnangan]({{ homePageURL }}).

Please log in using the following password:
```
{{ password }}
```

## Note
Do not share the password and change the default password as soon as possible.
