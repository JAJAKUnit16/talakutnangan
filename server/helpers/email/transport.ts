import { createTransport } from "nodemailer"
import type { Transporter, TransportOptions, SentMessageInfo } from "nodemailer"

import { Environment } from "!/types"
import RequestEnvironment from "!/helpers/request_environment"
import convertMarkdownToHTML from "!/helpers/convert_markdown_to_html"
import specializeTemplateFile from "!/helpers/specialize_template_file"

export default class Transport {
	private static currentInstance: Transport

	static initialize(
		host: string,
		port: number,
		senderUsername: string,
		senderPassword: string
	): void {
		if (!this.currentInstance) {
			this.currentInstance = new Transport(
				host,
				port,
				senderUsername,
				senderPassword
			)
		}
	}

	static async sendMail(to: string, subject: string, emailTemplatePath: string, variables: object)
		: Promise<SentMessageInfo> {
		return await this.currentInstance.sendMail(to, subject, emailTemplatePath, variables)
	}

	private senderUser: string
	private transport: Transporter

	constructor(host: string, port: number, user: string, pass: string) {
		this.senderUser = user
		this.transport = createTransport(<TransportOptions>{
			host,
			port,
			auth: {
				type: "login",
				user,
				pass
			},
			streamTransport: RequestEnvironment.isOnTest
		})

		this.transport.verify((error, success) => {
			if (error) {
				console.error("There is a problem on connecting with the e-mail server.")
				console.error(`Error [${error.name}]: ${error.message}`)
			} else {
				console.log("Connected to the e-mail server")
			}
		})
	}

	async sendMail(to: string, subject: string, emailTemplatePath: string, variables: object)
		: Promise<SentMessageInfo> {
		const text = await specializeTemplateFile(`email/${emailTemplatePath}`, variables)
		const html = convertMarkdownToHTML(text)

		const from = this.senderUser
		const message = {
			from,
			to,
			envelope: {
				from,
				to
			},
			subject: subject,
			text,
			html
		}

		return await new Promise((resolve, reject) => {
			this.transport.sendMail(message, (error, info) => {
				if (error) {
					reject(error)
				} else {
					resolve(info)
				}
			})
		})
	}
}
