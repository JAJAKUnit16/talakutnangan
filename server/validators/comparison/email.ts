import validator from "validator"

import type {
	ValidationState,
	ValidationConstraints
} from "!/types/validation"

/**
 * Validator to check if data is the proper acronym of the other field
 */
export default async function(
	currentState: Promise<ValidationState>,
	constraints: ValidationConstraints
): Promise<ValidationState> {
	const state = await currentState

	if (state.maySkip) return state

	if (validator.isEmail(state.value)) {
		return state
	}

	const error = {
		"field": constraints.field,
		"friendlyName": constraints.friendlyName,
		"messageMaker": (field: string) => `Field "${field}" must be valid email.`
	}

	throw error
}
