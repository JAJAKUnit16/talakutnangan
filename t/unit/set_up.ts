import "reflect-metadata"
import Database from "./database"

beforeAll(async () => {
	await Database.create()
});

afterAll(async () => {
	await Database.destroy()
});

afterEach(async () => {
	await Database.clear()
});
