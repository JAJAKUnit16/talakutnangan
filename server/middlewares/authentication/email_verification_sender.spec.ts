import "~/set-ups/email.set_up"
import MockRequester from "~/set-ups/mock_requester"

import type { EmailRequest } from "!/types/dependent"

import Transport from "!/helpers/email/transport"

import EmailVerificationSender from "./email_verification_sender"

describe("Middleware: Email Verification Sender", () => {
	const requester  = new MockRequester<EmailRequest>()

	it("can send to single user", async () => {
		const sender = new EmailVerificationSender()
		requester.customizeRequest({
			protocol: "http",
			hostname: "localhost",
			emailsToContact: [ "sample@example.com" ]
		})

		await requester.runMiddleware(sender.intermediate.bind(sender))

		requester.expectSuccess()
		const previousMessages = Transport.consumePreviousMessages()
		expect(previousMessages).toHaveLength(1)
		expect(previousMessages[0]).toHaveProperty("message")
		expect(previousMessages[0]).toHaveProperty("message.subject", "Email Verification")
		expect(previousMessages[0].message.text).toMatchSnapshot()
		expect(previousMessages[0].message.html).toMatchSnapshot()
	})

	it("can send to multiple users", async () => {
		const sender = new EmailVerificationSender()
		requester.customizeRequest({
			protocol: "http",
			hostname: "localhost",
			emailsToContact: [ "sampleA@example.com", "sampleB@example.net" ]
		})

		await requester.runMiddleware(sender.intermediate.bind(sender))

		requester.expectSuccess()
		const previousMessages = Transport.consumePreviousMessages()
		expect(previousMessages).toHaveLength(2)
		expect(previousMessages[0]).toHaveProperty("message")
		expect(previousMessages[0]).toHaveProperty("message.subject", "Email Verification")
		expect(previousMessages[0].message.text).toMatchSnapshot()
		expect(previousMessages[0].message.html).toMatchSnapshot()
		expect(previousMessages[1].message.text).toMatchSnapshot()
		expect(previousMessages[1].message.html).toMatchSnapshot()
	})
})
