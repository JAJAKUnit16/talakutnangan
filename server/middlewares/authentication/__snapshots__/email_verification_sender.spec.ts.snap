// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middleware: Email Verification Sender can send to multiple users 1`] = `
"# Email Verification

Your e-mail (sampleA@example.com) has been registered in [Talakutnangan](http://localhost).

Please click the button below to verify:
<a class=\\"verify\\" href=\\"http://localhost/user/verify?=sampleA@example.com\\">Verify e-mail</a>

## Note
If the button above does not work properly, visit http://localhost/user/verify?=sampleA@example.com
"
`;

exports[`Middleware: Email Verification Sender can send to multiple users 2`] = `
"<h1 id=\\"emailverification\\">Email Verification</h1>
<p>Your e-mail (sampleA@example.com) has been registered in <a href=\\"http://localhost\\">Talakutnangan</a>.</p>
<p>Please click the button below to verify:
<a class=\\"verify\\" href=\\"http://localhost/user/verify?=sampleA@example.com\\">Verify e-mail</a></p>
<h2 id=\\"note\\">Note</h2>
<p>If the button above does not work properly, visit http://localhost/user/verify?=sampleA@example.com</p>"
`;

exports[`Middleware: Email Verification Sender can send to multiple users 3`] = `
"# Email Verification

Your e-mail (sampleB@example.net) has been registered in [Talakutnangan](http://localhost).

Please click the button below to verify:
<a class=\\"verify\\" href=\\"http://localhost/user/verify?=sampleB@example.net\\">Verify e-mail</a>

## Note
If the button above does not work properly, visit http://localhost/user/verify?=sampleB@example.net
"
`;

exports[`Middleware: Email Verification Sender can send to multiple users 4`] = `
"<h1 id=\\"emailverification\\">Email Verification</h1>
<p>Your e-mail (sampleB@example.net) has been registered in <a href=\\"http://localhost\\">Talakutnangan</a>.</p>
<p>Please click the button below to verify:
<a class=\\"verify\\" href=\\"http://localhost/user/verify?=sampleB@example.net\\">Verify e-mail</a></p>
<h2 id=\\"note\\">Note</h2>
<p>If the button above does not work properly, visit http://localhost/user/verify?=sampleB@example.net</p>"
`;

exports[`Middleware: Email Verification Sender can send to single user 1`] = `
"# Email Verification

Your e-mail (sample@example.com) has been registered in [Talakutnangan](http://localhost).

Please click the button below to verify:
<a class=\\"verify\\" href=\\"http://localhost/user/verify?=sample@example.com\\">Verify e-mail</a>

## Note
If the button above does not work properly, visit http://localhost/user/verify?=sample@example.com
"
`;

exports[`Middleware: Email Verification Sender can send to single user 2`] = `
"<h1 id=\\"emailverification\\">Email Verification</h1>
<p>Your e-mail (sample@example.com) has been registered in <a href=\\"http://localhost\\">Talakutnangan</a>.</p>
<p>Please click the button below to verify:
<a class=\\"verify\\" href=\\"http://localhost/user/verify?=sample@example.com\\">Verify e-mail</a></p>
<h2 id=\\"note\\">Note</h2>
<p>If the button above does not work properly, visit http://localhost/user/verify?=sample@example.com</p>"
`;
